hpx/parallel/util/detail/chunk_size.hpp

```diff
@@ -15,50 +15,93 @@
 namespace hpx { namespace parallel { namespace util { namespace detail
 {
     ///////////////////////////////////////////////////////////////////////////
-    template <typename F, typename Future, typename FwdIter>
-        // requires traits::is_future<Future>
-    void add_ready_future(std::vector<Future>& workitems,
-        F && f, FwdIter first, std::size_t count)
-    {
-        workitems.push_back(hpx::make_ready_future(f(first, count)));
-    }
-
-    template <typename F, typename FwdIter>
-    void add_ready_future(std::vector<hpx::future<void> >& workitems,
-        F && f, FwdIter first, std::size_t count)
-    {
-        f(first, count);
-        workitems.push_back(hpx::make_ready_future());
-    }
 
-    template <typename F, typename FwdIter>
-    void add_ready_future(std::vector<hpx::shared_future<void> >& workitems,
-        F && f, FwdIter first, std::size_t count)
-    {
-        f(first, count);
-        workitems.push_back(hpx::make_ready_future());
-    }
-
-    // estimate a chunk size based on number of cores used
-    template <typename Future, typename F1, typename FwdIter>
+    // estimate a chunk size by running a quick benchmark
+    template <typename ExPolicy, typename Future, typename F1, typename FwdIter>
         // requires traits::is_future<Future>
     std::size_t auto_chunk_size(
+        ExPolicy const& policy,
         std::vector<Future>& workitems,
         F1 && f1, FwdIter& first, std::size_t& count)
     {
-        std::size_t test_chunk_size = count / 100;
-        if (0 == test_chunk_size) return 0;
-
+        // get executor
+        threads::executor exec = policy.get_executor();
+        
+        // get number of cores available
+        std::size_t const cores = hpx::get_os_thread_count(exec);
+
+        // get benchmark size. best value found so far: count/3
+        std::size_t test_chunk_size = (std::max)(count / (3*cores), (std::size_t)1);
+        
+        // get target chunk time
+        // TODO implement, if requested
+        boost::chrono::nanoseconds desired_chunktime_ns = boost::chrono::nanoseconds(0);//policy.get_chunk_time();
+        
+        // If no chunktime is supplied, fall back to 64us * cores
+        if(desired_chunktime_ns.count() == 0)
+            // TODO subject to change, further measurements necessary
+            desired_chunktime_ns = boost::chrono::nanoseconds(200000 + cores * 25000);
+        
+        // make sure we have enough work left to actually run the benchmark
+        if(count < test_chunk_size * cores) return 0;
+        
+        // run the benchmark iterations
+        std::vector<Future> bench_futures(cores);
         boost::uint64_t t = hpx::util::high_resolution_clock::now();
```

Comment:
Why default construct a whole bunch of future instances which get replaced in the next step anyways? Wouldn't be reserving the appropriate amount of elements in a default constructed vector combined with push_back below a better solution?